### PR TITLE
[TBBAS-957] EvalValue is not converted to an expected type on assignment fix

### PR DESCRIPTION
**Pull request by Yarikk26**
## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] Pull request title reflects its content
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# Description:

Fix for *EvalValue is not working as expected when assigned to a non-EvalValue variable*

### CODE REVIEW

File: core/coreobjects/include/coreobjects/property_builder.h
Comment (JakaMohorkoDS) [reply core/coreobjects/include/coreobjects/property_builder.h:102]: Do you have any suggestions on how best to handle polymorphic properties in Python? I'd think that this could be solved with overloads, but, according to Yaroslav's reasearch, pybind does not support overloads with property definitions. For function calls that accept eg. `INumber` I think this could be resolved more easily, but not for properties (get/set methods).

> The first one is that I would expect additional classifiers like evalValue on the parameter level rather than on the function level.

Here I'm not 100% sure on what's implied - we define RTGen information either at the interface or function level. Eg. `templateType()` also binds to a specific parameter of a function via its name. Do you have an example of how the parameter-level classifier would look?

> The second strange thing is that I would expect evalValue in some other functions as well, but the functions that already accept IBaseObject already use the desired syntax, so evalValue was probably left out in many cases just because there was no immediate need for it?

Do you have any examples?

Note that there's really almost no other functions that accept eval values as function arguments. They are almost exclusively used for property metadata. One thing that's likely missing here is that the Property and PropertyBuilder factories could also accept EvalValue parameters.

Comment (JakaMohorkoDS) [reply core/coreobjects/include/coreobjects/property_builder.h:102]: The other thought that occurs to me here... Is that in the Python bindings, we're not really using the fact that this was marked as `evalValue`, but we're simply marking said function parameters as base object types. In that case, `evalValue` means more that the parameter is polymorphic. 

Though this is just me being nitpicky.

**Review by JakaMohorkoDS — APPROVED** [review]
Mojca's comments should be addressed before merging.

File: core/coreobjects/include/coreobjects/property_builder.h
Comment (mojca) [reply core/coreobjects/include/coreobjects/property_builder.h:102]: > Do you have any suggestions on how best to handle polymorphic properties in Python? I'd think that this could be solved with overloads, but, according to Yaroslav's reasearch, pybind does not support overloads with property definitions. For function calls that accept eg. `INumber` I think this could be resolved more easily, but not for properties (get/set methods).

Pybind11 generally supports overloads. I do imagine that it might get tricky with properties. You cannot have an overloaded getter (not specific to pybind11; that simply doesn't make any sense), and because you cannot define partial properties (like defining only the getter or only the setter part; this is something that might be a valid feature request for pybind11, even if there are workarounds), I can easily imagine that there is in fact no way to properly support overloads there.

One existing ticket
* https://github.com/pybind/pybind11/issues/191
is suggesting precisely what has been implemented here: accept `py::object` and cast inside the function.

> > The first one is that I would expect additional classifiers like evalValue on the parameter level rather than on the function level.
> 
> Here I'm not 100% sure on what's implied - we define RTGen information either at the interface or function level. Eg. `templateType()` also binds to a specific parameter of a function via its name. Do you have an example of how the parameter-level classifier would look?

It would be nicer to be able to write:

```c++
    // [returnSelf]
    /*!
     * @brief Sets the Unit of the Property.
     * @param[evalValue] unit The Unit of the Property.
     */
    virtual ErrCode INTERFACE_FUNC getUnit(IUnit** unit) = 0;
```
instead of
```c++
    // [returnSelf, evalValue(unit)]
    /*!
     * @brief Sets the Unit of the Property.
     * @param unit The Unit of the Property.
     */
    virtual ErrCode INTERFACE_FUNC getUnit(IUnit** unit) = 0;
```
or something along those lines. But of course the above code cannot work without additional tricks in place. I'm not super familiar with doxygen, but I assume that the syntax after `@param` is strictly defined by doxygen syntax, while the comment above `// [returnSelf]` is only parsed by rtgen and generally ignored by doxygen.

> > The second strange thing is that I would expect evalValue in some other functions as well, but the functions that already accept IBaseObject already use the desired syntax, so evalValue was probably left out in many cases just because there was no immediate need for it?
> 
> Do you have any examples?

`setMinValue` has `evalValue` specified, while `setDefaultValue` does not. Is `setDefaultValue` really different in that respect, or is it just because "`setDefaultValue` already accepts `IBaseObject`, so we don't need to worry about adding that weird `evalValue` classifier there".

File: shared/tools/RTGen/src/project/RTGen.Python/Generators/PythonGenerator.cs
Comment (mojca) [inline shared/tools/RTGen/src/project/RTGen.Python/Generators/PythonGenerator.cs:109]: Just a random thought. Would it be acceptable/desirable to turn this into something along the lines of
```suggestion
                else if (argument.IsPolymorphic)
```
where `IsPolymorphic` would be implemented inside `Argument.cs` and would return exactly that result? (I leave it up to others to suggest a slightly better name.)

@Unsttopabull?

**Review by mojca — COMMENTED** [review]
After some discussion with @JakaMohorkoDS, here are my thoughts.

* The generated code in Python should most likely remain exactly the same as it is now (after your changes).
* I don't particularly like the way parameters are listed, but I don't see a much better way (we could have another keyword like `hasPolymorphicSetter`, but this would be less generic and we would need another change in the future in case we need to add a second parameter, so this doesn't make sense).
* (I'm still slightly confused that `setDefaultValue` doesn't have this parameter.)
* But it might help to pick a more descriptive name, though I also don't have a supper nice suggestion. My current understanding is that this parameter basically means "treat this parameter as polymorphic because the function can accept either INumber (or whatever else was declared here) or IEvalValue". The way it is implemented (= named) right now suggests to me that "this parameter **IS** eval value", which isn't exactly true.

Can we come up with a descriptive yet a much shorter version of "needsToBePolymorphicBecauseItCouldAlsoBeEvalValue"? If we don't have a good proposal, I'm also fine if we accept this as it is.

**Review by JakaMohorkoDS — APPROVED** [review]

**Pull request merged**
